### ISSUE TITLE
Find in Article (Alt-F) not working

### ISSUE THREAD

**Issue by lizy14** [kiwix/kiwix-js-windows#65]
I'm with latest version from Microsoft Store (0.9.9.81 Beta as shown in About), Windows 10 1809.

I've tried this on multiple articles from multiple ZIMs, but it won't work at all.

![image](https://user-images.githubusercontent.com/9985286/49661912-5f398680-fa85-11e8-8c36-9f3662f5ba3a.png)

**Comment by Jaifroid** [comment]
@lizy14 Thanks for the report, I'll investigate!

**Comment by Jaifroid** [comment]
@lizy14 I can't reproduce this problem on my system, with a fresh install of Kiwix JS from the Store. However, I don't yet have W10 1809, my machine has not yet received the update. I'd be very surprised if something specific in 1809 introduced an incompatibility in EdgeHTML with the JS code that does the in-page search, though anything is possible.

On Monday, I aim to release a major new update to Kiwix JS that supports TEDx video playback and "download" of epub ebooks from Project Gutenberg ZIMs. It doesn't have direct changes to the Ctrl/Alt-F in-page search code, but the update might reset something not working in your install. When you get the new update, let me know if the issue persists.

Alternatively, you could try uninstalling Kiwix JS (right-click the app icon and Uninstall) and then re-installing it from the Store. You won't have to re-download your ZIM files, they won't be touched by the uninstall, but you will have to pick their folder or the specific file again from the Config page. Or wait for the update after Monday, and we'll investigate further if the issue persists.

**Comment by Jaifroid** [comment]
@lizy14 , Kiwix JS version 0.9.9.85 Beta is now available in the Store. Could you please try it and see if your issue is fixed? Please make sure you see the splash screen like below after update (or check the version number in About). If your issue persists, please let me know which ZIM in particular you use, so I can try to reproduce. On the built-in Ray Charles test ZIM, I am not seeing any problem, nor with others I've tested. Many thanks.

![image](https://user-images.githubusercontent.com/4304337/49732281-fd8c3e80-fc75-11e8-9b54-7d6df4b4f6ea.png)

**Comment by lizy14** [comment]
I've updated to 0.9.9.85 Beta, and my issue persists. 
By accident, I realized that it relates to how quickly I type. 

![screenshot](https://user-images.githubusercontent.com/9985286/49747229-eaad5480-fcdd-11e8-996f-19a1d170bc21.gif)

However, for searching Chinese phrases, I still can't get it to work.

![image](https://user-images.githubusercontent.com/9985286/49747397-542d6300-fcde-11e8-8ed2-7f553f5003b0.png)

**Comment by Jaifroid** [comment]
@lizy14 Good sleuthing! That makes sense, and provides me with a good clue as to where the problem lies. I'll get back to you.

**Comment by Jaifroid** [comment]
@lizy14 I believe commit https://github.com/kiwix/kiwix-js-windows/commit/fa190c94afe28011b0c57851684501285dfc0e4e fixes this issue. There was a timeout set on the `keyup` event of the search box, designed to prevent race conditions. Removing the timeout does not seem to cause any race condition, and it appears to solve this issue. The fact that the issue occurred only with fast typing more or less confirms that the timeout was the issue.

This fix will make it into the next release of Kiwix JS Windows, probably next week. If you would like to have the fix before then, or would like to test, then you can install the "unreleased" package for Kiwix JS version 9.9.86 by following instructions [here](https://github.com/kiwix/kiwix-js-windows/tree/master-dev/AppPackages) - the package to sideload is the appxupload bundle `KiwixWebApp_0.9.986.0_AnyCPU...`. You may need to put your device into Developer Mode. However, as I assume the bug is non-critical, I would recommend waiting till the next Store release.

After the next Store release I'll contact you again to check that the issue is resolved before closing this.

**Comment by lizy14** [comment]
Thanks.  fa190c9 fixes searching 'piano' typing fast. 

However, searching for Chinese phrases is still not working. 

As far as I can see, the issue is with the call to `addAccents` at [www/js/lib/util.js:446](https://github.com/kiwix/kiwix-js-windows/blob/ee575655f3455d0041a7c735e7764f6453727c23/www/js/lib/util.js#L446) conflicting with  the call to `convertCharStr2jEsc` at [www/js/lib/util.js:578](https://github.com/kiwix/kiwix-js-windows/blob/ee575655f3455d0041a7c735e7764f6453727c23/www/js/lib/util.js#L578), for example,  

`"维基"` 
-> `"\\u7EF4\\u57FA"` 
-> `"\\[uùûüú]7[eèéêë]F4\\[uùûüú]57F([aàâäá]|ae)"`

**Comment by lizy14** [comment]
In addition, the default `setMatchType('left')` at [www/js/app/js:316](https://github.com/kiwix/kiwix-js-windows/blob/ee575655f3455d0041a7c735e7764f6453727c23/www/js/app.js#L316) does not work for Chinese (and many other languages: they are left-to-right, but there is no space between words).

**Comment by Jaifroid** [comment]
Ok, thanks.. As I have no experience with Chinese and as you obviously are familiar with code, feel free to offer a pull request. If you can't, I'll take look when I'm back, as I'm currently travelling, back mid week.

**Comment by Jaifroid** [comment]
OK, thanks! I'll test it with European languages when I'm back mid-week.

**Comment by kelson42** [comment]
@lizy14 @Jaifroid I enjoy reading about your collaboration on this ticket. Congrats to both of you!

**Comment by Jaifroid** [comment]
The fix for this issue is now in version 0.9.9.87 beta, both in Kiwix JS and WikiMed apps in the Microsoft Store. It will be incorporated into the Wikivoyage app next week.

**Comment by Jaifroid** [comment]
Hi @lizy14 . I've had to re-open this issue because the removal of the timeout on the Ctrl-F/Alt-F search field was causing a race condition in long articles (e.g. "Paris.html" in Wikipedia English), that sometimes makes the app completely unresponsive. This is caused by the fact that the app starts to search through thousands of instances of a single letter, anywhere in a word (in languages with words).

So, I've made a more robust timeout that prevents the search code being activated immediately on every letter press, but unlike the previous timeout code (that you removed), doesn't swallow letters or fail to register some of them.

I've also added in another form of protection, which is that if a user types ASCII letters with a space at start or end, then the search type is set to "left" (letters are matched at the beginning of words in left-to-right languages). However, if a user types a string of Chinese (or other non-ASCII) characters, then the search type remains "open", to support languages that don't have the concept of a typographical "word".

It's not perfect. If a user mixes Chinese and ASCII letters and also types a space, then the search will be set to the "left" type. However, it should work OK for simple searches of Chinese-only characters.

I'll keep this issue open to note when this fix goes into production. But if you want to take a look, it's commit https://github.com/kiwix/kiwix-js-windows/commit/e75784c987405d0b9107f170ac7c3ec5b772606a in the master-dev branch.
